redfish-core/lib/assembly.hpp

```diff
@@ -18,6 +21,94 @@
 namespace redfish
 {
 
+constexpr std::array<std::string_view, 9> chassisAssemblyInterfaces = {
+    "xyz.openbmc_project.Inventory.Item.Vrm",
+    "xyz.openbmc_project.Inventory.Item.Tpm",
+    "xyz.openbmc_project.Inventory.Item.Panel",
+    "xyz.openbmc_project.Inventory.Item.Battery",
+    "xyz.openbmc_project.Inventory.Item.DiskBackplane",
+    "xyz.openbmc_project.Inventory.Item.Board",
+    "xyz.openbmc_project.Inventory.Item.Connector",
+    "xyz.openbmc_project.Inventory.Item.Drive",
+    "xyz.openbmc_project.Inventory.Item.Board.Motherboard"};
+
+inline void doGetAssociatedChassisAssembly(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& chassisPath,
+    std::function<void(const std::vector<std::string>& assemblyList)>&&
+        callback)
+{
+    BMCWEB_LOG_DEBUG("Get associated chassis assembly");
+
+    sdbusplus::message::object_path endpointPath{chassisPath};
+    endpointPath /= "assembly";
+
+    dbus::utility::getAssociatedSubTreePaths(
+        endpointPath,
+        sdbusplus::message::object_path("/xyz/openbmc_project/inventory"), 0,
+        chassisAssemblyInterfaces,
+        [asyncResp, chassisPath, callback{callback}](
+            const boost::system::error_code& ec,
+            const dbus::utility::MapperGetSubTreePathsResponse& subtreePaths) {
+        if (ec)
+        {
+            if (ec.value() != EBADR)
+            {
+                BMCWEB_LOG_ERROR(
+                    "DBUS response error for getAssociatedSubTreePaths {}",
+                    ec.value());
+                messages::internalError(asyncResp->res);
+                return;
+            }
+            callback(std::vector<std::string>());
```

Comment:
Is the goal here to allow the callback to handle the error? Would it be clearer to handle the error here?

Maybe a comment here would be helpful.

Reply:
Yes, it is also a valid approach, but I've chosen to process here in case of the dbus error.
One reason why I chose this approach is like how `getValidChassisPath()` is done.
(If I choose to pass `ec` to caller, the caller handling may get crowed but not for the case of chassisPath).
Add a comment - 
```
            // Pass the empty assemblyList to caller
            callback(std::vector<std::string>());
```